pyat/test_matlab/test_cmp_physics.py

```diff
@@ -1,39 +1,134 @@
-from at import physics
+import math
+from at import physics, get_s_pos, uint32_refpts
+from scipy.constants import speed_of_light
 import matlab
 import numpy
 import pytest
 
 
+def _py_data(ml_data):
+    """Convert a Matlab vector to a numpy vector"""
+    return numpy.squeeze(numpy.asarray(ml_data))
+
+
+def _ml_refs(refpts, nelems):
+    """Convert refpoints to Matlab"""
+    uintrefs = uint32_refpts(refpts, nelems)
+    return matlab.double([ref+1 for ref in uintrefs])
+
+
 @pytest.mark.parametrize('dp', (0, 1e-8, 1e-7, 1e-6))
-@pytest.mark.parametrize('refpts', ([1], [1, 2, 3]))
+@pytest.mark.parametrize('refpts', (0, [0, 1, 2, -1], None, '-'))
 def test_find_orbit4(engine, ml_lattice, py_lattice, dp, refpts):
-    # Matlab call
-    ml_refpts = (matlab.double([]) if refpts is None else
-                 matlab.double(list(r + 1 for r in refpts)))
-    ml_orbit4 = engine.findorbit4(ml_lattice, dp, ml_refpts)
-    py_ml_orbit4 = numpy.asarray(ml_orbit4)
+
+    def ml_convert(orbit):
+        return numpy.rollaxis(numpy.asarray(orbit), -1)
+
+    nelems = len(py_lattice)
+    if isinstance(refpts, str):
+        refpts = range(nelems+1)
 
     # Python call
-    orbs, py_orbit4 = physics.find_orbit4(py_lattice, dp, refpts)
+    py_orb4, py_orbit4 = physics.find_orbit4(py_lattice, dp, refpts)
+    ml_orbit4, ml_orb4 = engine.findorbit4(ml_lattice, dp, _ml_refs(refpts, nelems), nargout=2)
 
-    numpy.testing.assert_almost_equal(py_ml_orbit4, py_orbit4[:4,:])
+    if refpts is None:
+        # test initial orbit
+        numpy.testing.assert_almost_equal(py_orb4, _py_data(ml_orb4), decimal=8)
+    else:
+        # test refpoints
+        numpy.testing.assert_almost_equal(py_orbit4[:, :4], ml_convert(ml_orbit4), decimal=8)
 
 
 @pytest.mark.parametrize('dp', (0.0, 1e-8, 1e-7, 1e-6))
-@pytest.mark.parametrize('refpts', ([1], [1, 2, 3], [145]))
+@pytest.mark.parametrize('refpts', (0, [0, 1, 2, -1], None, '-'))
```

Comment:
What do these `'-'` arguments mean?

Reply:
I am looking for an argument meaning "all refpoints up to the end", but the lattice length is unknown. `None` cannot be used, it already means "no refpoint". So I chose "any str argument means all refpoints", and I convert it immediately to `range(len(lattice)+1)`.
There are probably more elegant ways to do this, if a fixture can give both the lattice and its length, but can we postpone this to a next update ?